src/main/java/com/google/cloud/dataproc/templates/pubsub/PubSubToGCS.java

```diff
@@ -0,0 +1,227 @@
+    gcsBucketName = getProperties().getProperty(PUBSUB_GCS_BUCKET_NAME);
+    batchSize = Integer.parseInt(getProperties().getProperty(PUBSUB_GCS_BATCH_SIZE_PROP));
+    outputDataFormat = getProperties().getProperty(PUBSUB_GCS_OUTPUT_DATA_FORMAT);
+  }
+
+  @Override
+  public void runTemplate() {
+    if (StringUtils.isAllBlank(inputProjectID)
+        || StringUtils.isAllBlank(pubsubInputSubscription)
+        || StringUtils.isAllBlank(outputProjectID)
+        || StringUtils.isAllBlank(gcsBucketName)
+        || StringUtils.isAllBlank(outputDataFormat)) {
+      LOGGER.error(
+          "{},{},{},{},{} are required parameter. ",
+          PUBSUB_GCS_INPUT_PROJECT_ID_PROP,
+          PUBSUB_GCS_INPUT_SUBSCRIPTION_PROP,
+          PUBSUB_GCS_OUTPUT_PROJECT_ID_PROP,
+          PUBSUB_GCS_BUCKET_NAME,
+          PUBSUB_GCS_OUTPUT_DATA_FORMAT);
+      throw new IllegalArgumentException(
+          "Required parameters for PubSubToGCS not passed. "
+              + "Set mandatory parameter for PubSubToGCS template "
+              + "in resources/conf/template.properties file.");
+    }
+
+    JavaStreamingContext jsc = null;
+    LOGGER.info(
+        "Starting PubSub to GCS spark job with following parameters:"
+            + "1. {}:{}"
+            + "2. {}:{}"
+            + "3. {}:{}"
+            + "4. {},{}"
+            + "5. {},{}"
+            + "6. {},{}"
+            + "7. {},{}"
+            + "8. {},{}"
+            + "9. {},{}",
+        PUBSUB_GCS_INPUT_PROJECT_ID_PROP,
+        inputProjectID,
+        PUBSUB_GCS_INPUT_SUBSCRIPTION_PROP,
+        pubsubInputSubscription,
+        PUBSUB_GCS_TIMEOUT_MS_PROP,
+        timeoutMs,
+        PUBSUB_GCS_STREAMING_DURATION_SECONDS_PROP,
+        streamingDuration,
+        PUBSUB_GCS_TOTAL_RECEIVERS_PROP,
+        totalReceivers,
+        PUBSUB_GCS_OUTPUT_PROJECT_ID_PROP,
+        outputProjectID,
+        PUBSUB_GCS_BUCKET_NAME,
+        gcsBucketName,
+        PUBSUB_GCS_BATCH_SIZE_PROP,
+        batchSize,
+        PUBSUB_GCS_OUTPUT_DATA_FORMAT,
+        outputDataFormat);
+
+    try {
+      SparkConf sparkConf = new SparkConf().setAppName("PubSubToGCS Dataproc Job");
+      jsc = new JavaStreamingContext(sparkConf, Seconds.apply(streamingDuration));
+
+      JavaDStream<SparkPubsubMessage> stream = null;
+      for (int i = 0; i < totalReceivers; i += 1) {
+        JavaDStream<SparkPubsubMessage> pubSubReciever =
+            PubsubUtils.createStream(
+                jsc,
+                inputProjectID,
+                pubsubInputSubscription,
+                new SparkGCPCredentials.Builder().build(),
+                StorageLevel.MEMORY_AND_DISK_SER());
+        if (stream == null) {
+          stream = pubSubReciever;
+        } else {
+          stream = stream.union(pubSubReciever);
+        }
+      }
+
+      LOGGER.info("Writing data to output GCS Bucket: {}", gcsBucketName);
+      Storage storage = StorageOptions.getDefaultInstance().getService();
```

Comment:
Why is the code getting the storage bucket. 
Spark can do it automatically.
Refer: https://github.com/GoogleCloudPlatform/dataproc-templates/blob/main/src/main/java/com/google/cloud/dataproc/templates/databases/SpannerToGCS.java#L74

Reply:
@shashank-google , The SpannerToGCS uses Dataframe approach. I was following the same approach used in PubsubToBigQuery template. I could not find any code samples for reading PubSub data using Spark Dataframes. Please let me know if we have any references for it.